android/src/main/java/com/amplitude/android/utilities/Session.kt

```diff
@@ -0,0 +1,130 @@
+package com.amplitude.android.utilities
+
+import com.amplitude.android.Amplitude
+import com.amplitude.android.Configuration
+import com.amplitude.core.State
+import com.amplitude.core.Storage
+import com.amplitude.core.events.BaseEvent
+import java.util.concurrent.atomic.AtomicLong
+
+class Session(
+    private var configuration: Configuration,
+    private var storage: Storage? = null,
+    private var state: State? = null,
+) {
+    companion object {
+        const val EMPTY_SESSION_ID = -1L
+        const val DEFAULT_LAST_EVENT_TIME = -1L
+        const val DEFAULT_LAST_EVENT_ID = 0L
+    }
+
+    private val _sessionId = AtomicLong(EMPTY_SESSION_ID)
+    private val _lastEventId = AtomicLong(DEFAULT_LAST_EVENT_ID)
+    private val _lastEventTime = AtomicLong(DEFAULT_LAST_EVENT_TIME)
+
+    val sessionId: Long
+        get() {
+            return _sessionId.get()
+        }
+
+    internal var lastEventId: Long = DEFAULT_LAST_EVENT_ID
+        get() = _lastEventId.get()
+    var lastEventTime: Long = DEFAULT_LAST_EVENT_TIME
+        get() = _lastEventTime.get()
+
+    init {
+        loadFromStorage()
+    }
+
+    private fun loadFromStorage() {
+        _sessionId.set(storage?.read(Storage.Constants.PREVIOUS_SESSION_ID)?.toLongOrNull() ?: EMPTY_SESSION_ID)
+        _lastEventId.set(storage?.read(Storage.Constants.LAST_EVENT_ID)?.toLongOrNull() ?: DEFAULT_LAST_EVENT_ID)
+        _lastEventTime.set(storage?.read(Storage.Constants.LAST_EVENT_TIME)?.toLongOrNull() ?: DEFAULT_LAST_EVENT_TIME)
+    }
+
+    /**
+     * startNewSessionIfNeeded
+     *
+     * @param timestamp By default this is used as both `sessionId` and `timestamp`
+     * @param sessionId If set, this is used as `sessionId`
+     */
+    suspend fun startNewSessionIfNeeded(timestamp: Long, sessionId: Long? = null): Iterable<BaseEvent>? {
```

Comment:
How do we handle extending sessions when the app is in foreground?

Reply:
Sessions stay active as long as the app is in the foreground, and will set lastEventTime onExitForeground. Some of that logic is done while processing events in the timeline.

Reply:
Ah cool. It's a little different compared to how swift does it. Swift had a flag for inForeground.